backend/azure/client.go

```diff
@@ -0,0 +1,189 @@
+	if err != nil {
+		return err
+	}
+
+	containerURL := azblob.NewContainerURL(*URL, a.pipeline)
+	blobURL := containerURL.NewBlockBlobURL(utils.RemoveLeadingSlash(file.Path()))
+	_, err = blobURL.SetMetadata(context.Background(), metadata, azblob.BlobAccessConditions{}, azblob.ClientProvidedKeyOptions{})
+	return err
+}
+
+// Download returns an io.ReadCloser for the given vfs.File
+func (a *ClientImpl) Download(file vfs.File) (io.ReadCloser, error) {
+	URL, err := url.Parse(file.Location().(*Location).ContainerURL())
+	if err != nil {
+		return nil, err
+	}
+
+	containerURL := azblob.NewContainerURL(*URL, a.pipeline)
+	blobURL := containerURL.NewBlockBlobURL(utils.RemoveLeadingSlash(file.Path()))
+	get, err := blobURL.Download(context.Background(), 0, 0, azblob.BlobAccessConditions{}, false, azblob.ClientProvidedKeyOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return get.Body(azblob.RetryReaderOptions{}), nil
+}
+
+// Copy copies srcFile to the destination tgtFile within Azure Blob Storage.  Note that in the case where we get
+// encoded spaces in the file name (i.e. %20) the '%' must be encoded or the copy command will return a not found
+// error.
+func (a *ClientImpl) Copy(srcFile vfs.File, tgtFile vfs.File) error {
+	// Can't use url.PathEscape here since that will escape everything (even the directory separators)
+	srcURL, err := url.Parse(strings.Replace(srcFile.URI(), "%", "%25", -1))
+	if err != nil {
+		return err
+	}
+
+	tgtURL, err := url.Parse(tgtFile.Location().(*Location).ContainerURL())
+	if err != nil {
+		return err
+	}
+
+	containerURL := azblob.NewContainerURL(*tgtURL, a.pipeline)
+	blobURL := containerURL.NewBlockBlobURL(utils.RemoveLeadingSlash(tgtFile.Path()))
+	ctx := context.Background()
+	resp, err := blobURL.StartCopyFromURL(ctx, *srcURL, azblob.Metadata{}, azblob.ModifiedAccessConditions{}, azblob.BlobAccessConditions{}, azblob.DefaultAccessTier, nil)
+	if err != nil {
+		return err
+	}
+
+	for resp.CopyStatus() == azblob.CopyStatusPending {
+		time.Sleep(2 * time.Second)
+	}
+
+	if resp.CopyStatus() == azblob.CopyStatusSuccess {
+		return nil
+	}
+
+	return fmt.Errorf("Copy failed.  ERROR[%s]", resp.ErrorCode())
+}
+
+// List will return a listing of the contents of the given location.  Each item in the list will contain the full key
+// as specified by the azure blob (incliding the virtual 'path').
+func (a *ClientImpl) List(l vfs.Location) ([]string, error) {
+	URL, err := url.Parse(l.(*Location).ContainerURL())
+	if err != nil {
+		return []string{}, err
+	}
+
+	containerURL := azblob.NewContainerURL(*URL, a.pipeline)
+	ctx := context.Background()
+	var list []string
+	for marker := (azblob.Marker{}); marker.NotDone(); {
+		listBlob, err := containerURL.ListBlobsHierarchySegment(ctx, marker, "/", azblob.ListBlobsSegmentOptions{Prefix: utils.RemoveLeadingSlash(l.Path())})
+		if err != nil {
+			log.Fatal("Unable to list blobs.  error: " + err.Error())
```

Comment:
I found this while working on another ticket.  I have fixed it in that branch.  Pretty sure we don't want this in there.  :)